File: problems/prob11/prob11.clj
(use '[clojure.string :only [split split-lines]])

(defn strings-to-ints [strlist]
  "Convert the strlist to a vector of ints"
  (vec (map #(Integer/valueOf %) strlist)))

(def grid
  (vec (map #(strings-to-ints (split % #" "))
            (split-lines (slurp "grid.txt")))))

(defn diag-products [grid x y]
  "Return the main and secondary diagonals of the 4x4 matrix with (x,y) as top left corner."
  (let [cell #((grid %1) %2)] ;cell accessor function
    (list
     (reduce * [(cell x y) (cell (+ x 1) (+ y 1)) (cell (+ x 2) (+ y 2)) (cell (+ x 3) (+ y 3))])
     (reduce * [(cell (+ x 3) y) (cell (+ x 2) (+ y 1)) (cell (+ x 1) (+ y 2)) (cell x (+ y 3))]))))

(defn matrix-products [grid x y]
  "Return all the possible, non-redundant 4-element products of the 4x4 matrix with (x,y) as top-left corner."
  (let [limit (- (count grid) 4)]
    (concat
     (if (<= y limit)
       (list (apply * (->> (grid x) (drop y) (take 4))))) ; add row product
     (if (<= x limit)
       (list (apply * (->> (map #(nth % y) grid) (drop x) (take 4))))) ; add col product
     (if (and (<= y limit) (<= x limit))
       (diag-products grid x y)))))

(defn max-product [grid]
  "Calculate the products starting from each cell of the matrix"
  (let [num-range (range (count grid))]
    (->> (for [x num-range y num-range] (matrix-products grid x y))
         (apply concat)
         (apply max))))

(println (max-product grid))

File: problems/prob11/prob11.py
grid = []
with open('grid.txt') as f:
    for line in f:
        grid.append([int(x) for x in line.split()])

def mul(l):
    """ Return the product of the elements in l """
    if len(l) == 0: return 0
    product = 1
    for el in l: 
        product *= el
    return product

def products(grid, x, y):
    prods = []

    # row products
    for r in range(x, x+4):
        prods.append(mul([grid[r][c] for c in range(y, y+4)]))

    # column products
    for c in range(y, y+4):
        prods.append(mul([grid[r][c] for r in range(x, x+4)]))

    # diagonal products
    prods.append(mul([grid[x][y], grid[x+1][y+1], grid[x+2][y+2], grid[x+3][y+3]]))
    prods.append(mul([grid[x+3][y], grid[x+2][y+1], grid[x+1][y+2], grid[x][y+3]]))

    return prods

from itertools import chain
print(max(list(chain(
    *[products(grid, x, y) for x in range(len(grid)-4) for y in range(len(grid)-4)]))))
